### PR TITLE
Support mutable lambdas in sharded::invoke_on_others

### PR DESCRIPTION
**Pull request by xemul**
Currently only sharded::invoke_on_all() can call mutable lambdas, the ..._on_others fails to compile. This PR fixes it and adds test for more invoke_on_... compilation options.

### CODE REVIEW

File: tests/unit/sharded_test.cc [hunk 1]
```diff
@@ -201,3 +201,32 @@ SEASTAR_THREAD_TEST_CASE(invoke_on_all_sharded_arg) {
     srv.stop().get();
     arg.stop().get();
 }
+
+SEASTAR_THREAD_TEST_CASE(invoke_on_modifiers) {
+    class checker {
+    public:
+        future<> fn(int a) {
```
[the remark this answers is not in the document]
Comment (tchaikov) [reply tests/unit/sharded_test.cc:208]: also, might want to rename `fn` to a name which can be distinguished from its non-const cousin. for instance, "access()" versus "mutate()".

and,  this function is never used in this test. is this expected?

when the compiler looks up for a function based on its arguments,  the function which is resolved to, is not dependent to if the enclosing lambda has the `mutable` specifier. it depends on, well, the arguments' types. 

in this case, if the `this` argument which is implicitly passed to the argument is `const`, the const version is used. see https://godbolt.org/z/M77v3Yzbc.

so, i think it'd be better to use different names for `const` function and for non-const function when testing the mutable lambda and non-mutable lambda.

File: tests/unit/sharded_test.cc [hunk 2]
```diff
@@ -201,3 +201,32 @@ SEASTAR_THREAD_TEST_CASE(invoke_on_all_sharded_arg) {
...
+    };
+
+    seastar::sharded<checker> srv;
+    srv.start().get();
+    int a = 42;
+
+    srv.invoke_on_all([a] (checker& s) { return s.fn(a); }).get();
+    srv.invoke_on_all([a] (checker& s) mutable { return s.fn(a); }).get();
+    srv.invoke_on_others([a] (checker& s) { return s.fn(a); }).get();
+    srv.invoke_on_others([a] (checker& s) mutable ->future<> { return s.fn(a); }).get();
```
Comment (tchaikov) [inline tests/unit/sharded_test.cc:223]: how is this test case related to the fix?

File: tests/unit/sharded_test.cc [hunk 3]
```diff
@@ -201,3 +201,32 @@ SEASTAR_THREAD_TEST_CASE(invoke_on_all_sharded_arg) {
...
+        future<> fn(int a) const {
+            return make_ready_future<>();
+        }
+    };
+
+    seastar::sharded<checker> srv;
+    srv.start().get();
+    int a = 42;
+
+    srv.invoke_on_all([a] (checker& s) { return s.fn(a); }).get();
```
Comment (tchaikov) [inline tests/unit/sharded_test.cc:220]: i'd suggest add a test case which pass a `const &` to the lambda, like
```c++
srv.invoke_on_all([a] (const checker& s) { return s.fn(a); }).get();
```

**Review by tchaikov — COMMENTED** [review]
left a couple comments.

**Comment by avikivity** [comment]
How can mutable lambdas work? All shards will try to mutate them concurrently.

File: tests/unit/sharded_test.cc (hunk 3 above)
Comment (xemul) [reply tests/unit/sharded_test.cc:220]: It's below, and it's commented out, because it doesn't work currently: #2278

File: tests/unit/sharded_test.cc (hunk 2 above)
Comment (xemul) [reply tests/unit/sharded_test.cc:223]: The lambda is marked `mutable`, this is what the fix fixes -- if you remove first patch from this PR and try to compile _this very line_ will fail compilation

File: tests/unit/sharded_test.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (xemul) [reply tests/unit/sharded_test.cc:208]: > nit, IMHO, the argument is not quite relevant to the problem we are trying to resolve here.

I need to have calling lambda with capture (below), so I capture `int a`. In order not to have unused capture, I pass it here, so the argument is indeed unused. I can plug it the other way, if you prefer

[the remark this answers is not in the document]
Comment (xemul) [reply tests/unit/sharded_test.cc:208]: > and, this function is never used in this test. is this expected?

The non-const fn is in use. The `const` one isn't, but it's left "for future", see #2278. I can remove it for now

**Comment by xemul** [comment]
> How can mutable lambdas work? All shards will try to mutate them concurrently.

Code can try to std::move() an argument from it, e.g. [scylla/transport/server.cc](https://github.com/scylladb/scylladb/blob/79d0711c7e37e34ee28cc4a871da0c3e86bcc559/transport/server.cc#L1065-L1073)

```c++
    return parallel_for_each(cpus.begin(), cpus.end(), [this, query, cpu_id, &client_state] (unsigned int c) mutable {
        if (c != cpu_id) {
            return smp::submit_to(c, [this, query, &client_state] () mutable {
                return _server._query_processor.local().prepare(std::move(query), client_state, false).discard_result();
            });
        } else {
            return make_ready_future<>();
        }
    })
```

this call to `parallel_for_each(cpus, ...)` is in fact `_server._query_processor.invoke_on_others(...)` one, which cannot compile

**Comment by avikivity** [comment]
> > How can mutable lambdas work? All shards will try to mutate them concurrently.
> 
> Code can try to std::move() an argument from it, e.g. [scylla/transport/server.cc](https://github.com/scylladb/scylladb/blob/79d0711c7e37e34ee28cc4a871da0c3e86bcc559/transport/server.cc#L1065-L1073)
> 
> ```c++
>     return parallel_for_each(cpus.begin(), cpus.end(), [this, query, cpu_id, &client_state] (unsigned int c) mutable {
>         if (c != cpu_id) {
>             return smp::submit_to(c, [this, query, &client_state] () mutable {
>                 return _server._query_processor.local().prepare(std::move(query), client_state, false).discard_result();
>             });
>         } else {
>             return make_ready_future<>();
>         }
>     })
> ```
> 
> this call to `parallel_for_each(cpus, ...)` is in fact `_server._query_processor.invoke_on_others(...)` one, which cannot compile

How can this work? The first iteration will clear `query` and the second one will work from an empty variable.

File: tests/unit/sharded_test.cc (hunk 3 above)
Comment (tchaikov) [reply tests/unit/sharded_test.cc:220]: it's not below. could you please read the comment again? please.

File: tests/unit/sharded_test.cc (hunk 2 above)
Comment (tchaikov) [reply tests/unit/sharded_test.cc:223]: i meant, what's the point of `->future<>`?
how is 
```c++
srv.invoke_on_all([a] (checker& s) mutable { return s.fn(a); }).get();
```
different from
```
srv.invoke_on_all([a] (checker& s) mutable ->future<> { return s.fn(a); }).get();
```
?

File: tests/unit/sharded_test.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (tchaikov) [reply tests/unit/sharded_test.cc:208]: > I need to have calling lambda with capture (below)

this is why i am confused. what makes it important to capture something in this context? i assume you want to test something. but what is it?

[the remark this answers is not in the document]
Comment (tchaikov) [reply tests/unit/sharded_test.cc:208]: > > and, this function is never used in this test. is this expected?
> 
> The non-const fn is in use. The `const` one isn't, but it's left "for future", see #2278. I can remove it for now

could you please re-read my comment? and probably the linked godbolt example, if you haven't. i commented out the non-const fn, and the sample compiles just fine.

**Comment by xemul** [comment]
> How can this work? The first iteration will clear `query` and the second one will work from an empty variable.

The inner `smp::submit_to()` copies the lambda capture block including `query` string.
So does the invoke_on_(all|others)(), they both try to copy it.

**Comment by xemul** [comment]
Or maybe the don't copy it, but in that case I cannot explain why this code fails to compile

```c++
SEASTAR_THREAD_TEST_CASE(lambda_copies_stuff) {
    class foo {
    public:
        foo() {}
        foo(const foo&) = delete;
        foo(foo&&) = delete;
        void fn() const {
            fmt::print("{}\n", this_shard_id());
        }
    };
    foo f;
    smp::submit_to(1, [f] {
        f.fn();
    }).get();
}
```

```
/home/xemul/src/seastar/tests/unit/sharded_test.cc:245:23: error: use of deleted function ‘lambda_copies_stuff::do_run_test_case() const::foo::foo(const lambda_copies_stuff::do_run_test_case() const::foo&)’
  245 |     smp::submit_to(1, [f] {
      |                       ^~~~~
  246 |         f.fn();
      |         ~~~~~~~        
  247 |     }).get();
      |     ~                  
/home/xemul/src/seastar/tests/unit/sharded_test.cc:238:9: note: declared here
  238 |         foo(const foo&) = delete;
      |         ^~~
```

File: tests/unit/sharded_test.cc (hunk 2 above)
Comment (xemul) [reply tests/unit/sharded_test.cc:223]: Ah, the `-> future<>` can be omitted indeed

File: tests/unit/sharded_test.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (xemul) [reply tests/unit/sharded_test.cc:208]: > what makes it important to capture something in this context? 

Because lambda without capture, even being mutable, compiles just fine without the fix. I _do_ need non-empty capture here. And if I use `[a]` capture, but don't do anything with `a` inside lambda, compiler warns me about unused lambda capture. So I pass it as argument to `foo::fn(int)`

[the remark this answers is not in the document]
Comment (xemul) [reply tests/unit/sharded_test.cc:208]: > could you please re-read my comment? and probably the linked godbolt example, if you haven't. i commented out the non-const fn, and the sample compiles just fine.

I don't see how it's related. My code is

```
   srv.invoke_on_all([a] (checker& s) { return s.fn(a); });
```

`srv` is non-const, `checker& s` argument is non-const either, `s.fn(a)` calls for non-const `fn()` (if I remove const overload, it compiles just fine)

File: tests/unit/sharded_test.cc (hunk 3 above)
Comment (xemul) [reply tests/unit/sharded_test.cc:220]: OK, I missed that you suggest to have `const checker&`, but _not_ `const_cast<const ...>(srv)` part. What for?

File: tests/unit/sharded_test.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (xemul) [reply tests/unit/sharded_test.cc:208]: Maybe this would explain my intent better. I don't try to test if the lambda calls const fn or non const fn depending on whether it's mutable or not. My point is that invoke_on_other() doesn't work with mutable lambdas at all, it just fails to compile like this

```
/home/xemul/src/seastar/include/seastar/core/future.hh:2074:11: error: no type named ‘type’ in ‘struct std::invoke_result<const invoke_on_modifiers::do_run_test_case() const::<lambda(invoke_on_modifiers::do_run_test_case() const::checker&)>&, invoke_on_modifiers::do_run_test_case() const::checker&>’
 2074 |     using futurator = futurize<std::invoke_result_t<Func, Args&&...>>;
      |           ^~~~~~~~~
```

[the remark this answers is not in the document]
Comment (xemul) [reply tests/unit/sharded_test.cc:208]: I updated the PR with less code in the test. Please, check, hope it demonstrates the problem in a cleaner way

[the remark this answers is not in the document]
Comment (tchaikov) [reply tests/unit/sharded_test.cc:208]: > > what makes it important to capture something in this context?
> 
> Because lambda without capture, even being mutable, compiles just fine without the fix. I _do_ need non-empty capture here. And if I use `[a]` capture, but don't do anything with `a` inside lambda, compiler warns me about unused lambda capture. So I pass it as argument to `foo::fn(int)`

thanks. TIL.

File: tests/unit/sharded_test.cc (hunk 3 above)
Comment (tchaikov) [reply tests/unit/sharded_test.cc:217]: hmm, just for the sake of completeness. but it's not important.

**Review by tchaikov — APPROVED** [review]
lgtm

**Comment by xemul** [comment]
@avikivity , what's you decision here -- is it wrong to have mutable lambdas in invoke_on_...'s or is it worth it (so this PR needs to be merged)?

**Comment by avikivity** [comment]
> @avikivity , what's you decision here -- is it wrong to have mutable lambdas in invoke_on_...'s or is it worth it (so this PR needs to be merged)?

I think it's even documented that invoke_on works on a copy.
